dcos_containers input plugin

philipnrmn opened the pull request:
This PR adds the dcos_containers input plugin. 

The plugin gets container data from the mesos operator API on every call to Gather(). 

Resolves [DCOS_OSS-3716](https://jira.mesosphere.com/browse/DCOS_OSS-3716)

philipnrmn committed:
Add README for dcos_containers

philipnrmn committed:
Import mock mesos server

philipnrmn committed:
Add empty test fixture

philipnrmn committed:
Test that gather does not error

philipnrmn committed:
Add skeleton for dcos_containers plugin

philipnrmn committed:
Add getContainers method

philipnrmn committed:
Add measurement struct

philipnrmn committed:
Add test method for timestamps

philipnrmn committed:
Add setIfNotNil method and test

philipnrmn committed:
Flesh out test for Gather

philipnrmn committed:
Add a test case with container data

philipnrmn committed:
Add methods to get fields, tags and timestamp

philipnrmn committed:
Add testcase for blkio stats

philipnrmn committed:
Add methods to get blkio fields

philipnrmn committed:
Add dcos_containers to all.go

janisz commented on plugins/inputs/dcos_containers/README.md:
```diff
@@ -0,0 +1,90 @@
+  ## The URL of the mesos agent
+  mesos_agent_url = "http://localhost:5051"
+  ## The period after which requests to mesos agent should time out
+  timeout = 10s
```
I'm not toml expert but github higlight `s` I think it should be wrpped in `"` as regular string.
https://github.com/toml-lang/toml#user-content-spec

philipnrmn replied:
👍 thanks, I fixed this

janisz commented on plugins/inputs/dcos_containers/dcos_containers.go:
```diff
@@ -0,0 +1,310 @@
+// measurement is a combination of fields and tags specific to those fields
+type measurement struct {
+	name   string
+	fields map[string]interface{}
```
Why `fields` values are `interface{}` it looks like we want to keep there only numeric values is this correct? Why not simply use `float64`

philipnrmn replied:
Thanks for looking over my PR! `fields` are values returned from getters on the [mesos proto](https://github.com/mesos/mesos-go/blob/master/api/v1/lib/mesos.pb.go). Those values can be `uint64`, `uint32`, or `float64`. 

Since they are fed into the `telegraf.Accumulator` which expects `map[string]interface{}`: https://github.com/dcos/telegraf/blob/1.7.2-dcos/accumulator.go#L14-L17, it makes sense to store them as `map[string]interface{}` in the interim.

philipnrmn committed:
Fix formatting of time duration in toml

philipnrmn commented on plugins/inputs/dcos_containers/dcos_containers.go:
```diff
@@ -0,0 +1,310 @@
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
```
@branden do you think imports like this should be from the `influxdata` or the `dcos` org? I've used `influxdata` in this PR but maybe that's not right...

branden replied:
I've been using `influxdata` since otherwise we'd have to rewrite a bunch of the imports in this repo to point to `dcos`. It looks like it's typical for forks of golang projects to use the original project's package name: http://blog.campoy.cat/2014/03/github-and-go-forking-pull-requests-and.html

philipnrmn commented on plugins/inputs/dcos_containers/dcos_containers.go:
```diff
@@ -0,0 +1,310 @@
+	warnIfNotSet(setIfNotNil(disk.fields, "limit_bytes", rs.GetDiskLimitBytes))
+	warnIfNotSet(setIfNotNil(disk.fields, "used_bytes", rs.GetDiskUsedBytes))
+
+	// TODO (philipnrmn) *rs.DiskStatistics for per-volume stats
```
These TODOs are for a post 1.12 world, rs.DiskStatistics is currently not used in dcos-metrics.

branden commented on plugins/inputs/dcos_containers/cmd/gen.go:
```diff
@@ -0,0 +1,56 @@
+//
+// You should run 'go generate' every time you change one of the json files in
+// the testdata directory, and commit both the changed json file and the
+// changed binary file.
```
It may be helpful to add a `testdata/README.md` that contains this notice. I could see myself editing stuff under there without thinking to look at this file.

branden commented on plugins/inputs/dcos_containers/testdata/empty/README.md:
```diff
@@ -0,0 +1,5 @@
+
+- Given that no task are running on the cluster
+- When container metrics are retrieved
+- Then no container metrics should be presnet
```
Typo: `present`

branden commented on plugins/inputs/dcos_containers/mock_mesos_server.go:
```diff
@@ -0,0 +1,63 @@
+	})
+	server := httptest.NewServer(router)
+
+	return server, server.Close
```
Why return `server.Close` here, instead of letting the caller get it from the `server`?

philipnrmn replied:
I originally had a dedicated teardown method, I removed it.

branden commented on plugins/inputs/dcos_containers/dcos_containers.go:
```diff
@@ -102,6 +139,87 @@ func processResponse(resp mesos.Response, t agent.Response_Type) (agent.Response
+
+// cTags extracts relevant metadata from a Container object as a map of tags
+func cTags(c agent.Response_GetContainers_Container) map[string]string {
+	return map[string]string{"container_id": c.ContainerID.Value}
```
Should this be `dcos_container_id`?

branden commented on plugins/inputs/dcos_containers/dcos_containers.go:
```diff
@@ -0,0 +1,310 @@
+
+// cTags extracts relevant metadata from a Container object as a map of tags
+func cTags(c agent.Response_GetContainers_Container) map[string]string {
+	return map[string]string{"container_id": c.ContainerID.Value}
```
Should this be `dcos_container_id`? I've been assuming that the tags we add for DC/OS-specific values (e.g. cluster ID, node role) would be prefixed with `dcos_`.

philipnrmn replied:
Tricky one. I have been using `container_id` since container IDs are mesos-level rather than DC/OS-level concepts. But then this whole plugin is `dcos_containers` so maybe that's wrong. Do you have strong feelings either way?

branden replied:
My only concern is whether the metrics API output can identify the metrics that come from this input, so that it can pass them along to the HTTP producer.  `container_id` seems a like a generic enough tag that looking just for that risks creating false positives, though realistically it probably won't cause any problems. I'll see if I can find other hints to use, and fall back to just looking for a `container_id` tag if I can't.

branden reviewed (approved):
This looks great, I especially like the tests. I have a few questions and nits inlined but this LGTM.

philipnrmn committed:
Improve README files for test fixtures

philipnrmn committed:
Remove superfluous teardown method

janisz commented on plugins/inputs/dcos_containers/dcos_containers.go:
```diff
@@ -0,0 +1,310 @@
+func processResponse(resp mesos.Response, t agent.Response_Type) (agent.Response, error) {
+	var r agent.Response
+	defer func() {
+		if resp != nil {
```
Do we need this condition?

philipnrmn committed:
Test for the presence of ResourceStatistics

ResourceStatistics can be empty in some cases, which could cause a segfault

philipnrmn committed:
Add info prefix to log

Telegraf convention is to use I! for info and E! for error

philipnrmn committed:
Use internal.Duration for configurable timeouts

internal.Duration has an unmarshal-toml method which allows it to be read from a
config file as a string, eg "10s"

philipnrmn committed:
Fix go vet issues

philipnrmn committed:
Bump test timeout to 100ms

josepedro commented on plugins/inputs/dcos_containers/dcos_containers.go:
```diff
@@ -0,0 +1,322 @@
+	resp, err := cli.Send(ctx, calls.NonStreaming(calls.GetContainers()))
+	if err != nil {
+		return nil, err
+	}
```
You could put 
`if err != nil {
            return nil, err
}`
on a separated function and call it on 108 and 112 lines.

philipnrmn replied:
Wouldn't I have to test its response and return only if it was nil?

josepedro replied:
Yep its true
